File: app/src/main/java/io/ipoli/android/quest/widgets/AgendaWidgetProvider.java
package io.ipoli.android.quest.widgets;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.widget.RemoteViews;
import android.widget.Toast;

import com.squareup.otto.Bus;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import javax.inject.Inject;

import io.ipoli.android.Constants;
import io.ipoli.android.MainActivity;
import io.ipoli.android.R;
import io.ipoli.android.app.App;
import io.ipoli.android.app.activities.QuickAddActivity;
import io.ipoli.android.app.events.EventSource;
import io.ipoli.android.player.persistence.PlayerPersistenceService;
import io.ipoli.android.quest.activities.QuestActivity;
import io.ipoli.android.quest.events.AgendaWidgetDisabledEvent;
import io.ipoli.android.quest.events.AgendaWidgetEnabledEvent;
import io.ipoli.android.quest.events.CompleteQuestRequestEvent;
import io.ipoli.android.quest.persistence.QuestPersistenceService;

public class AgendaWidgetProvider extends AppWidgetProvider {

    private static final String WIDGET_SHOW_QUEST_ACTION = "io.ipoli.android.intent.actions.WIDGET_SHOW_QUEST_ACTION";
    private static final String WIDGET_ADD_QUEST_ACTION = "io.ipoli.android.intent.actions.WIDGET_ADD_QUEST_ACTION";

    public static final String QUEST_ACTION_EXTRA_KEY = "quest_action";

    public static final int QUEST_ACTION_VIEW = 1;
    public static final int QUEST_ACTION_COMPLETE = 2;

    @Inject
    Bus eventBus;

    @Inject
    QuestPersistenceService questPersistenceService;

    @Inject
    PlayerPersistenceService playerPersistenceService;

    @Override
    public void onEnabled(Context context) {
        super.onEnabled(context);
        eventBus.post(new AgendaWidgetEnabledEvent());
    }

    @Override
    public void onDisabled(Context context) {
        eventBus.post(new AgendaWidgetDisabledEvent());
        super.onDisabled(context);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        App.getAppComponent(context).inject(this);
        if (WIDGET_SHOW_QUEST_ACTION.equals(intent.getAction())) {
            String questId = intent.getStringExtra(Constants.QUEST_ID_EXTRA_KEY);

            int questAction = intent.getIntExtra(QUEST_ACTION_EXTRA_KEY, 0);
            if (questAction == QUEST_ACTION_VIEW) {
                onViewQuest(context, questId);
            } else if (questAction == QUEST_ACTION_COMPLETE) {
                onQuestComplete(questId);
            }
        } else if (WIDGET_ADD_QUEST_ACTION.equals(intent.getAction())) {
            if (!App.hasPlayer()) {
                Toast.makeText(context, R.string.widget_quick_add_validation, Toast.LENGTH_LONG).show();
                return;
            }
            Intent i = new Intent(context, QuickAddActivity.class);
            i.putExtra(Constants.QUICK_ADD_ADDITIONAL_TEXT, " " + context.getString(R.string.today).toLowerCase());
            i.setAction(Long.toString(System.currentTimeMillis()));
            i.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(i);
        }
        super.onReceive(context, intent);
    }

    private void onViewQuest(Context context, String questId) {
        Intent i = new Intent(context, QuestActivity.class);
        i.putExtra(Constants.QUEST_ID_EXTRA_KEY, questId);
        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(i);
    }

    private void onQuestComplete(String questId) {
        questPersistenceService.findById(questId, quest -> {
            if (quest == null) {
                return;
            }
            eventBus.post(new CompleteQuestRequestEvent(quest, EventSource.WIDGET));
        });
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {

        for (int appWidgetId : appWidgetIds) {
            RemoteViews rv = new RemoteViews(context.getPackageName(), R.layout.widget_agenda);
            setupHeader(context, rv);
            Intent widgetServiceIntent = createWidgetServiceIntent(context, appWidgetId);
            setupAgenda(widgetServiceIntent, context, appWidgetId, rv);
            setupEmptyView(context, rv);
            updateWidget(appWidgetManager, appWidgetId, rv);
        }
        super.onUpdate(context, appWidgetManager, appWidgetIds);
    }

    @NonNull
    private Intent createWidgetServiceIntent(Context context, int appWidgetId) {
        Intent intent = new Intent(context, AgendaWidgetService.class);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        intent.setData(Uri.parse(intent.toUri(Intent.URI_INTENT_SCHEME)));
        return intent;
    }

    private void updateWidget(AppWidgetManager appWidgetManager, int appWidgetId, RemoteViews rv) {
        appWidgetManager.updateAppWidget(appWidgetId, rv);
    }

    @NonNull
    private Intent setupAgenda(Intent widgetServiceIntent, Context context, int appWidgetId, RemoteViews rv) {
        rv.setRemoteAdapter(R.id.widget_agenda_list, widgetServiceIntent);
        Intent templateIntent = new Intent(context, AgendaWidgetProvider.class);
        templateIntent.setAction(AgendaWidgetProvider.WIDGET_SHOW_QUEST_ACTION);
        templateIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        widgetServiceIntent.setData(Uri.parse(widgetServiceIntent.toUri(Intent.URI_INTENT_SCHEME)));
        PendingIntent templatePendingIntent = PendingIntent.getBroadcast(context, 0, templateIntent,
                PendingIntent.FLAG_UPDATE_CURRENT);
        rv.setPendingIntentTemplate(R.id.widget_agenda_list, templatePendingIntent);
        return widgetServiceIntent;
    }

    private void setupHeader(Context context, RemoteViews rv) {
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat(context.getString(R.string.widget_agenda_title_format), Locale.getDefault());
        rv.setTextViewText(R.id.widget_agenda_title, simpleDateFormat.format(new Date()));
        setupHeadClickListener(context, rv);
        setupAddClickListener(context, rv);
    }

    private void setupHeadClickListener(Context context, RemoteViews rv) {
        Intent startAppIntent = new Intent(context, MainActivity.class);
        PendingIntent startAppPendingIntent = PendingIntent.getActivity(context, 0, startAppIntent, PendingIntent.FLAG_CANCEL_CURRENT);
        rv.setOnClickPendingIntent(R.id.widget_agenda_header, startAppPendingIntent);
    }

    private void setupAddClickListener(Context context, RemoteViews rv) {
        rv.setOnClickPendingIntent(R.id.widget_agenda_add, getQuickAddPendingIntent(context));
    }

    private PendingIntent getQuickAddPendingIntent(Context context) {
        Intent intent = new Intent(context, AgendaWidgetProvider.class);
        intent.setAction(AgendaWidgetProvider.WIDGET_ADD_QUEST_ACTION);
        return PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private void setupEmptyView(Context context, RemoteViews rv) {
        rv.setOnClickPendingIntent(R.id.widget_agenda_empty, getQuickAddPendingIntent(context));
        rv.setEmptyView(R.id.widget_agenda_list, R.id.widget_agenda_empty);
    }
}

File: app/src/main/java/io/ipoli/android/quest/widget/agenda/AgendaWidgetProvider.kt
package io.ipoli.android.quest.widget.agenda

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews
import io.ipoli.android.Constants
import io.ipoli.android.MyPoliApp
import io.ipoli.android.R
import io.ipoli.android.common.IntentUtil
import io.ipoli.android.common.di.BackgroundModule
import io.ipoli.android.common.text.CalendarFormatter
import io.ipoli.android.quest.receiver.CompleteQuestReceiver
import kotlinx.coroutines.experimental.Dispatchers
import kotlinx.coroutines.experimental.GlobalScope
import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.withContext
import org.threeten.bp.LocalDate
import org.threeten.bp.format.TextStyle
import space.traversal.kapsule.Injects
import space.traversal.kapsule.inject
import space.traversal.kapsule.required
import java.util.*


/**
 * Created by Venelin Valkov <venelin@mypoli.fun>
 * on 02/10/2018.
 */
class AgendaWidgetProvider : AppWidgetProvider(), Injects<BackgroundModule> {

    private val playerRepository by required { playerRepository }

    companion object {
        const val WIDGET_QUEST_LIST_ACTION =
            "mypoli.android.intent.actions.WIDGET_QUEST_LIST_ACTION"

        const val QUEST_ACTION_EXTRA_KEY = "quest_action"

        const val QUEST_ACTION_VIEW = 1
        const val QUEST_ACTION_COMPLETE = 2
    }

    override fun onReceive(context: Context, intent: Intent) {

        if (WIDGET_QUEST_LIST_ACTION == intent.action) {
            val questId = intent.getStringExtra(Constants.QUEST_ID_EXTRA_KEY)

            val questAction = intent.getIntExtra(QUEST_ACTION_EXTRA_KEY, 0)

            if (questAction == QUEST_ACTION_VIEW) {
                context.startActivity(IntentUtil.showTimer(questId, context))

            } else if (questAction == QUEST_ACTION_COMPLETE) {
                onCompleteQuest(context, questId)

            } else throw IllegalArgumentException("Unknown agenda widget quest list action $questAction")
        }

        super.onReceive(context, intent)
    }

    private fun onCompleteQuest(context: Context, questId: String) {
        val i = Intent(context, CompleteQuestReceiver::class.java)
        i.putExtra(Constants.QUEST_ID_EXTRA_KEY, questId)
        context.sendBroadcast(i)
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        inject(MyPoliApp.backgroundModule(context))
        val calendarFormatter = CalendarFormatter(MyPoliApp.instance)

        val today = LocalDate.now()
        val dayOfWeek = today.dayOfWeek.getDisplayName(
            TextStyle.FULL, Locale.getDefault()
        )
        val date = calendarFormatter.dateWithoutYear(today)

        appWidgetIds.forEach {
            GlobalScope.launch(Dispatchers.IO) {
                val player = playerRepository.find()

                withContext(Dispatchers.Main) {

                    val rv = RemoteViews(context.packageName, R.layout.widget_agenda)

                    rv.setTextViewText(R.id.widgetDayOfWeek, dayOfWeek)
                    rv.setTextViewText(R.id.widgetDate, date)

                    rv.setOnClickPendingIntent(
                        R.id.widgetAgendaHeader,
                        createStartAppIntent(context)
                    )
                    rv.setOnClickPendingIntent(R.id.widgetAgendaPet, createShowPetIntent(context))
                    rv.setOnClickPendingIntent(R.id.widgetAgendaAdd, createQuickAddIntent(context))

                    if(player != null && player.isDead) {
                        rv.setViewVisibility(R.id.widgetAgendaPlayerDiedContainer, View.VISIBLE)
                        rv.setViewVisibility(R.id.widgetAgendaList, View.GONE)
                        rv.setViewVisibility(R.id.widgetAgendaEmpty, View.GONE)
                        rv.setViewVisibility(R.id.widgetAgendaPet, View.GONE)
                        rv.setViewVisibility(R.id.widgetAgendaAdd, View.GONE)

                        rv.setOnClickPendingIntent(
                            R.id.widgetAgendaRevive,
                            createStartAppIntent(context)
                        )
                    } else {

                        rv.setViewVisibility(R.id.widgetAgendaPlayerDiedContainer, View.GONE)
                        rv.setViewVisibility(R.id.widgetAgendaList, View.VISIBLE)
                        rv.setViewVisibility(R.id.widgetAgendaEmpty, View.VISIBLE)
                        rv.setViewVisibility(R.id.widgetAgendaPet, View.VISIBLE)
                        rv.setViewVisibility(R.id.widgetAgendaAdd, View.VISIBLE)

                        rv.setRemoteAdapter(
                            R.id.widgetAgendaList,
                            createQuestListIntent(context, it)
                        )

                        rv.setPendingIntentTemplate(
                            R.id.widgetAgendaList,
                            createQuestClickIntent(context, it)
                        )

                        rv.setEmptyView(R.id.widgetAgendaList, R.id.widgetAgendaEmpty)
                    }

                    appWidgetManager.notifyAppWidgetViewDataChanged(it, R.id.widgetDayOfWeek)
                    appWidgetManager.notifyAppWidgetViewDataChanged(it, R.id.widgetDate)

                    appWidgetManager.notifyAppWidgetViewDataChanged(it, R.id.widgetAgendaList)
                    appWidgetManager.updateAppWidget(it, rv)
                }
            }
        }

        super.onUpdate(context, appWidgetManager, appWidgetIds)

    }

    private fun createQuickAddIntent(context: Context) =
        IntentUtil.getActivityPendingIntent(context, IntentUtil.showQuickAdd(context))

    private fun createShowPetIntent(context: Context) =
        IntentUtil.getActivityPendingIntent(context, IntentUtil.showPet(context))

    private fun createQuestListIntent(context: Context, widgetId: Int) =
        Intent(context, AgendaWidgetService::class.java).apply {
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
        }

    private fun createQuestClickIntent(context: Context, widgetId: Int): PendingIntent {
        val intent = Intent(context, AgendaWidgetProvider::class.java)
        intent.action = AgendaWidgetProvider.WIDGET_QUEST_LIST_ACTION
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
        return PendingIntent.getBroadcast(
            context,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT
        )
    }

    private fun createStartAppIntent(context: Context) =
        IntentUtil.getActivityPendingIntent(
            context,
            IntentUtil.startApp(context)
        )
}